formal_spec/nmt.qnt

```diff
@@ -0,0 +1,847 @@
+                        (acc, leaf_key) => 
+                        union(acc, Set(tree.leaves.get(leaf_key).namespaceId))                    
+                    ).oneOf()                 
+                    namespace_v' = namespaceId,
+                }
+            },
+            // unchanged variables
+            proof_v' = proof_v,                
+            corrupted' = corrupted,
+            corruption_type' = corruption_type,
+            corruption_diff' = corruption_diff
+            }
+    }
+
+    action generation = {
+        all{
+            // state preconditions
+            state_v == "generation",
+            // update state for the next step
+            state_v' = "final",
+            proof_v' = CreateProofNamespace(namespace_v, tree_v),
+            // unchanged variables
+            namespace_v' = namespace_v,
+            corrupted' = false,
+            tree_v' = tree_v,
+            leaves_v' = leaves_v,
+            corruption_type' = corruption_type,
+            corruption_diff' = corruption_diff            
+        }
+    }
+
+
+   
+
+
+
+    action final = {
+        all {
+            // state preconditions
+            state_v == "final",
+            //update state for the next step
+            state_v' = "requirements",
+            // resetting the state to initial values
+            namespace_v' = -1,
+            tree_v' = EMPTY_TREE,
+            proof_v' = EMPTY_PROOF,
+            corrupted' = false,
+            leaves_v' = EMPTY_LEAVES,
+            corruption_type' = "",
+            corruption_diff' = 
+            {
+                changed_start: -1,
+                changed_end: -1,
+                changed_namespace: -1,
+                changed_indices: []
+            }
+            }
+    }
+
+    action corruptProof = {
+        all{
+            // state preconditions
+            state_v == "final",
+            corrupted == false,
+
+            // corrupting the proof
+            any{
+                // corrupting the start value
+                all{
+                    proof_v.end != 1,
+                    namespace_v' = namespace_v,
+                    nondet new_start = oneOf(
+                        0.to(proof_v.end - 1).exclude(Set(proof_v.start))
+                        )
+                    all{
+                                        
+                        proof_v' = {start: new_start, end: proof_v.end, supporting_hashes: proof_v.supporting_hashes},                        
+                        corruption_type' = "start",                        
+                        corruption_diff' = corruption_diff.with("changed_start", new_start),
+                    }
+                },
+
+                // corrupting the end value
+                all{
+                    proof_v.start < proof_v.end - 1,
+                    namespace_v' = namespace_v,
+                    nondet new_end = oneOf(proof_v.start.to(proof_v.end - 1))
```

Comment:
Yes, you are right. And, similarly to the previous comment, we could add that kind of corruption as well.

A broader comment is, that if we see we want to add many different corruptions, then it probably makes sense to simply create data randomly instead of slightly perturbing correct proofs. The reason why I decided to go this way was because of a suspicion that slightly perturbed proofs could be closer to potentially unhandled edge cases.

Reply:
I agree that there are advantages and disadvantages to both randomly generated proofs and deliberately corrupted proofs.